diff -u -r xrick-021212-orig/include/game.h xrick-021212/include/game.h
--- xrick-021212-orig/include/game.h	2002-12-24 15:32:56.000000000 +0200
+++ xrick-021212/include/game.h	2011-11-29 13:14:33.454886155 +0200
@@ -28,7 +28,7 @@
 #define TRUE 1
 #define FALSE 0
 
-#define GAME_PERIOD 75
+#define GAME_PERIOD 50
 
 #define GAME_BOMBS_INIT 6
 #define GAME_BULLETS_INIT 6
diff -u -r xrick-021212-orig/include/system.h xrick-021212/include/system.h
--- xrick-021212-orig/include/system.h	2002-12-24 15:32:56.000000000 +0200
+++ xrick-021212/include/system.h	2011-11-29 13:05:45.970890984 +0200
@@ -76,7 +76,7 @@
 /*
  * video section
  */
-#define SYSVID_ZOOM 2
+#define SYSVID_ZOOM 1
 #define SYSVID_MAXZOOM 4
 #define SYSVID_WIDTH 320
 #define SYSVID_HEIGHT 200
Only in xrick-021212: Makefile.global
diff -u -r xrick-021212-orig/src/syssnd.c xrick-021212/src/syssnd.c
--- xrick-021212-orig/src/syssnd.c	2002-12-24 15:33:43.000000000 +0200
+++ xrick-021212/src/syssnd.c	2011-11-29 13:00:12.022894042 +0200
@@ -125,7 +125,7 @@
   }
 
   desired.freq = SYSSND_FREQ;
-  desired.format = AUDIO_U8;
+  desired.format = AUDIO_S8;
   desired.channels = SYSSND_CHANNELS;
   desired.samples = SYSSND_MIXSAMPLES;
   desired.callback = syssnd_callback;
diff -u -r xrick-021212-orig/src/system.c xrick-021212/src/system.c
--- xrick-021212-orig/src/system.c	2002-12-24 15:33:43.000000000 +0200
+++ xrick-021212/src/system.c	2011-11-29 13:54:55.738863977 +0200
@@ -20,6 +20,7 @@
 #include <signal.h>
 
 #include "system.h"
+#include <android/log.h>
 
 /*
  * Panic
@@ -42,6 +43,7 @@
 
   /* print message and die */
   printf("%s\npanic!\n", s);
+  __android_log_print(ANDROID_LOG_FATAL, "XRick", "Error: %s", s);
   exit(1);
 }
 
@@ -65,6 +67,7 @@
   vsprintf(s, msg, argptr);
   va_end(argptr);
   printf(s);
+  __android_log_print(ANDROID_LOG_INFO, "XRick", "%s", s);
 }
 
 /*
diff -u -r xrick-021212-orig/src/sysvid.c xrick-021212/src/sysvid.c
--- xrick-021212-orig/src/sysvid.c	2002-12-24 15:33:43.000000000 +0200
+++ xrick-021212/src/sysvid.c	2011-11-29 13:10:59.218888117 +0200
@@ -29,11 +29,11 @@
 
 static SDL_Color palette[256];
 static SDL_Surface *screen;
-static U32 videoFlags;
+static U32 videoFlags = SDL_SWSURFACE;
 
 static U8 zoom = SYSVID_ZOOM; /* actual zoom level */
-static U8 szoom = 0;  /* saved zoom level */
-static U8 fszoom = 0;  /* fullscreen zoom level */
+static U8 szoom = 1;  /* saved zoom level */
+static U8 fszoom = 1;  /* fullscreen zoom level */
 
 #include "img_icon.e"
 
@@ -221,7 +221,7 @@
   SDL_WM_SetIcon(s, NULL);
 
   /* video modes and screen */
-  videoFlags = SDL_HWSURFACE|SDL_HWPALETTE;
+  videoFlags = SDL_SWSURFACE;
   sysvid_chkvm();  /* check video modes */
   if (sysarg_args_zoom)
     zoom = sysarg_args_zoom;
@@ -315,12 +315,13 @@
     area.y = rects->y * zoom;
     area.h = rects->height * zoom;
     area.w = rects->width * zoom;
-    SDL_UpdateRects(screen, 1, &area);
+    /* SDL_UpdateRects(screen, 1, &area); */
 
     rects = rects->next;
   }
 
   SDL_UnlockSurface(screen);
+  SDL_Flip(screen);
 }
 
 
