From 421c7f351ded614e3c1eb0d10c079e5d26865bd5 Mon Sep 17 00:00:00 2001
From: Juanjo <juanjo.ng.83@gmail.com>
Date: Sat, 19 Oct 2013 11:48:29 +0000
Subject: [PATCH 006/249] Position of landscaping toolbar.

---
 src/terraform_gui.cpp | 10 +---------
 1 file changed, 1 insertion(+), 9 deletions(-)

diff --git a/src/terraform_gui.cpp b/src/terraform_gui.cpp
index 89e6204..2524b29 100644
--- a/src/terraform_gui.cpp
+++ b/src/terraform_gui.cpp
@@ -259,13 +259,6 @@ struct TerraformToolbarWindow : Window {
 		VpSelectTilesWithMethod(pt.x, pt.y, select_method);
 	}
 
-	virtual Point OnInitialPosition(int16 sm_width, int16 sm_height, int window_number)
-	{
-		Point pt = GetToolbarAlignedWindowPosition(sm_width);
-		pt.y += sm_height;
-		return pt;
-	}
-
 	virtual void OnPlaceMouseUp(ViewportPlaceMethod select_method, ViewportDragDropSelectionProcess select_proc, Point pt, TileIndex start_tile, TileIndex end_tile)
 	{
 		if (pt.x != -1) {
@@ -348,7 +341,7 @@ static const NWidgetPart _nested_terraform_widgets[] = {
 };
 
 static WindowDesc _terraform_desc(
-	WDP_MANUAL, "toolbar_landscape", 0, 0,
+	WDP_ALIGN_TOOLBAR, "toolbar_landscape", 0, 0,
 	WC_SCEN_LAND_GEN, WC_NONE,
 	WDF_CONSTRUCTION,
 	_nested_terraform_widgets, lengthof(_nested_terraform_widgets),
@@ -372,7 +365,6 @@ Window *ShowTerraformToolbar(Window *link)
 	DeleteWindowById(WC_SCEN_LAND_GEN, 0, true);
 	Window *w = AllocateWindowDescFront<TerraformToolbarWindow>(&_terraform_desc, 0);
 	/* Align the terraform toolbar under the main toolbar. */
-	w->top -= w->height;
 	w->SetDirty();
 	/* Put the linked toolbar to the left / right of it. */
 	link->left = w->left + (_current_text_dir == TD_RTL ? w->width : -link->width);
-- 
1.8.1.2

