From 199983be811c042a54b4079139d5521e9dcce9d5 Mon Sep 17 00:00:00 2001
From: Juanjo <juanjo.ng.83@gmail.com>
Date: Fri, 9 Aug 2013 09:33:12 +0000
Subject: [PATCH 172/249] Center yes/no buttons on query string.

---
 src/misc_gui.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/misc_gui.cpp b/src/misc_gui.cpp
index d771f92..2e54b4f 100644
--- a/src/misc_gui.cpp
+++ b/src/misc_gui.cpp
@@ -1155,9 +1155,13 @@ static const NWidgetPart _nested_query_widgets[] = {
 	EndContainer(),
 	NWidget(WWT_PANEL, COLOUR_RED), SetPIP(8, 15, 8),
 		NWidget(WWT_TEXT, COLOUR_RED, WID_Q_TEXT), SetMinimalSize(200, 12),
-		NWidget(NWID_HORIZONTAL, NC_EQUALSIZE), SetPIP(20, 29, 20),
-			NWidget(WWT_PUSHTXTBTN, COLOUR_YELLOW, WID_Q_NO), SetMinimalSize(71, 12), SetDataTip(STR_QUIT_NO, STR_NULL),
-			NWidget(WWT_PUSHTXTBTN, COLOUR_YELLOW, WID_Q_YES), SetMinimalSize(71, 12), SetDataTip(STR_QUIT_YES, STR_NULL),
+		NWidget(NWID_HORIZONTAL),
+			NWidget(WWT_EMPTY), SetFill(1, 1),
+			NWidget(NWID_HORIZONTAL, NC_EQUALSIZE), SetPIP(20, 29, 20),
+				NWidget(WWT_PUSHTXTBTN, COLOUR_YELLOW, WID_Q_NO), SetMinimalSize(71, 12), SetDataTip(STR_QUIT_NO, STR_NULL),
+				NWidget(WWT_PUSHTXTBTN, COLOUR_YELLOW, WID_Q_YES), SetMinimalSize(71, 12), SetDataTip(STR_QUIT_YES, STR_NULL),
+			EndContainer(),
+			NWidget(WWT_EMPTY), SetFill(1, 1),
 		EndContainer(),
 	EndContainer(),
 };
-- 
1.8.1.2

