From 0d255c000ba8913e3e8c6c48922d09071f5ba3e6 Mon Sep 17 00:00:00 2001
From: Juanjo <juanjo.ng.83@gmail.com>
Date: Sat, 10 Aug 2013 15:19:00 +0000
Subject: [PATCH 196/249] Center sprites on bridge picker window.

---
 src/bridge_gui.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/bridge_gui.cpp b/src/bridge_gui.cpp
index 61c477a..78e7ff8 100644
--- a/src/bridge_gui.cpp
+++ b/src/bridge_gui.cpp
@@ -231,7 +231,8 @@ public:
 					SetDParam(1, b->speed);
 					SetDParam(0, b->material);
 
-					DrawSprite(b->sprite, b->pal, r.left + WD_MATRIX_LEFT, y + this->resize.step_height - 1 - GetSpriteSize(b->sprite).height);
+					uint y_sprite = Center(y, this->resize.step_height, GetSpriteSize(b->sprite).height);
+					DrawSprite(b->sprite, b->pal, r.left + WD_MATRIX_LEFT, y_sprite);
 					DrawStringMultiLine(r.left + this->bridgetext_offset, r.right, y + 2, y + this->resize.step_height,
 							_game_mode == GM_EDITOR ? STR_SELECT_BRIDGE_SCENEDIT_INFO : STR_SELECT_BRIDGE_INFO);
 					y += this->resize.step_height;
-- 
1.8.1.2

