diff --git a/src/cdogs/files.h b/src/cdogs/files.h
index 2a231df..b42ce0d 100644
--- a/src/cdogs/files.h
+++ b/src/cdogs/files.h
@@ -81,6 +81,8 @@ ssize_t f_read(FILE *f, void *buf, size_t size);
 ssize_t f_read32(FILE *f, void *buf, size_t size);
 ssize_t f_read16(FILE *f, void *buf, size_t size);
 
+#undef swap32
+#undef swap16
 void swap32 (void *d);
 void swap16 (void *d);
 
diff --git a/src/json/json.c b/src/json/json.c
index fb928d4..be53ea8 100644
--- a/src/json/json.c
+++ b/src/json/json.c
@@ -26,6 +26,7 @@
 #include <assert.h>
 #include <memory.h>
 #include <sys/types.h>
+#include <limits.h>
 
 
 enum LEX_VALUE
diff --git a/CMakeLists.txt b/CMakeLists.txt
index c44844e..3d1170d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -95,7 +95,7 @@ else()
 	endif()
 endif()
 
-set(CDOGS_DATA_DIR "../")
+set(CDOGS_DATA_DIR "./")
 set(CDOGS_CFG_DIR ".cdogs/")
 SET(SOURCE_DIRECTORY ".")
 SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY ".")
