diff --git a/project/java/Globals.java b/project/java/Globals.java
index 04df3ed..bddaeb1 100644
--- a/project/java/Globals.java
+++ b/project/java/Globals.java
@@ -90,7 +90,7 @@ class Globals
 	public static int TrackballDampening = 0;
 	public static int AudioBufferConfig = 0;
 	public static boolean OptionalDataDownload[] = null;
-	public static int LeftClickMethod = Mouse.LEFT_CLICK_NORMAL;
+	public static int LeftClickMethod = Mouse.LEFT_CLICK_NEAR_CURSOR;
 	public static int LeftClickKey = KeyEvent.KEYCODE_DPAD_CENTER;
 	public static int LeftClickTimeout = 3;
 	public static int RightClickTimeout = 4;
@@ -103,7 +103,7 @@ class Globals
 	public static boolean RelativeMouseMovement = ForceRelativeMouseMode; // Laptop touchpad mode
 	public static int RelativeMouseMovementSpeed = 2;
 	public static int RelativeMouseMovementAccel = 0;
-	public static int ShowScreenUnderFinger = Mouse.ZOOM_NONE;
+	public static int ShowScreenUnderFinger = Mouse.ZOOM_MAGNIFIER;
 	public static boolean KeepAspectRatio = KeepAspectRatioDefaultSetting;
 	public static int ClickScreenPressure = 0;
 	public static int ClickScreenTouchspotSize = 0;
