diff -ru orig/milkytracker-0.90.85/configure milkytracker-0.90.85/configure
--- orig/milkytracker-0.90.85/configure	2010-01-01 19:03:37.000000000 -0500
+++ milkytracker-0.90.85/configure	2010-11-29 11:37:25.000000000 -0500
@@ -5431,95 +5431,9 @@
 fi
 
 pkg_failed=no
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for ZZIP" >&5
-$as_echo_n "checking for ZZIP... " >&6; }
-
-if test -n "$PKG_CONFIG"; then
-    if test -n "$ZZIP_CFLAGS"; then
-        pkg_cv_ZZIP_CFLAGS="$ZZIP_CFLAGS"
-    else
-        if test -n "$PKG_CONFIG" && \
-    { { $as_echo "$as_me:${as_lineno-$LINENO}: \$PKG_CONFIG --exists --print-errors \"zziplib >= 0.10.75\""; } >&5
-  ($PKG_CONFIG --exists --print-errors "zziplib >= 0.10.75") 2>&5
-  ac_status=$?
-  $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
-  test $ac_status = 0; }; then
-  pkg_cv_ZZIP_CFLAGS=`$PKG_CONFIG --cflags "zziplib >= 0.10.75" 2>/dev/null`
-else
-  pkg_failed=yes
-fi
-    fi
-else
-	pkg_failed=untried
-fi
-if test -n "$PKG_CONFIG"; then
-    if test -n "$ZZIP_LIBS"; then
-        pkg_cv_ZZIP_LIBS="$ZZIP_LIBS"
-    else
-        if test -n "$PKG_CONFIG" && \
-    { { $as_echo "$as_me:${as_lineno-$LINENO}: \$PKG_CONFIG --exists --print-errors \"zziplib >= 0.10.75\""; } >&5
-  ($PKG_CONFIG --exists --print-errors "zziplib >= 0.10.75") 2>&5
-  ac_status=$?
-  $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
-  test $ac_status = 0; }; then
-  pkg_cv_ZZIP_LIBS=`$PKG_CONFIG --libs "zziplib >= 0.10.75" 2>/dev/null`
-else
-  pkg_failed=yes
-fi
-    fi
-else
-	pkg_failed=untried
-fi
-
-
-
-if test $pkg_failed = yes; then
-
-if $PKG_CONFIG --atleast-pkgconfig-version 0.20; then
-        _pkg_short_errors_supported=yes
-else
-        _pkg_short_errors_supported=no
-fi
-        if test $_pkg_short_errors_supported = yes; then
-	        ZZIP_PKG_ERRORS=`$PKG_CONFIG --short-errors --errors-to-stdout --print-errors "zziplib >= 0.10.75"`
-        else
-	        ZZIP_PKG_ERRORS=`$PKG_CONFIG --errors-to-stdout --print-errors "zziplib >= 0.10.75"`
-        fi
-	# Put the nasty error message in config.log where it belongs
-	echo "$ZZIP_PKG_ERRORS" >&5
-
-	as_fn_error "Package requirements (zziplib >= 0.10.75) were not met:
 
 $ZZIP_PKG_ERRORS
 
-Consider adjusting the PKG_CONFIG_PATH environment variable if you
-installed software in a non-standard prefix.
-
-Alternatively, you may set the environment variables ZZIP_CFLAGS
-and ZZIP_LIBS to avoid the need to call pkg-config.
-See the pkg-config man page for more details.
-" "$LINENO" 5
-elif test $pkg_failed = untried; then
-	{ { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error "The pkg-config script could not be found or is too old.  Make sure it
-is in your PATH or set the PKG_CONFIG environment variable to the full
-path to pkg-config.
-
-Alternatively, you may set the environment variables ZZIP_CFLAGS
-and ZZIP_LIBS to avoid the need to call pkg-config.
-See the pkg-config man page for more details.
-
-To get pkg-config, see <http://pkg-config.freedesktop.org/>.
-See \`config.log' for more details." "$LINENO" 5; }
-else
-	ZZIP_CFLAGS=$pkg_cv_ZZIP_CFLAGS
-	ZZIP_LIBS=$pkg_cv_ZZIP_LIBS
-        { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-$as_echo "yes" >&6; }
-	:
-fi
-
 SDL_VERSION=1.2.0
 
 # Check whether --with-sdl-prefix was given.
@@ -5774,7 +5688,7 @@
 
   rm -f conf.sdltest
 
-CPPFLAGS="$CPPFLAGS -D__FORCE_SDL_AUDIO__"
+CPPFLAGS="$CPPFLAGS -D__FORCE_SDL_AUDIO__ -D__LOWRES__ -D__LOWRES__"
 
 # Check that SDL supports the X11 wm
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether SDL supports X11" >&5
Only in milkytracker-0.90.85: configure~
