diff -ru orig/vice-2.3/src/arch/sdl/uimenu.c vice-2.3/src/arch/sdl/uimenu.c
--- orig/vice-2.3/src/arch/sdl/uimenu.c	2011-02-19 11:33:41.000000000 -0500
+++ vice-2.3/src/arch/sdl/uimenu.c	2011-02-23 09:56:42.000000000 -0500
@@ -750,8 +750,9 @@
     if (vsid_mode && (sdl_vsid_state & SDL_VSID_ACTIVE)) {
         sdl_vsid_close();
     }
-
+#ifndef ANDROID
     SDL_EnableKeyRepeat(SDL_DEFAULT_REPEAT_DELAY, SDL_DEFAULT_REPEAT_INTERVAL);
+#endif
     sdl_menu_state = 1;
     ui_check_mouse_cursor();
 }
@@ -762,8 +763,9 @@
 
     sdl_menu_state = 0;
     ui_check_mouse_cursor();
+#ifndef ANDROID
     SDL_EnableKeyRepeat(0, 0);
-
+#endif
     /* Do not resume sound if in warp mode */
     resources_get_int("WarpMode", &warp_state);
     if (warp_state == 0) {

