--- simutrans/simmain.cc	2014-08-02 23:17:18.000000000 +0200
+++ simutrans/simmain.cc	2014-08-02 23:10:51.000000000 +0200
@@ -77,6 +77,9 @@
 #include "vehicle/simvehikel.h"
 #include "vehicle/simverkehr.h"
 
+#include <SDL.h>
+#include <SDL_screenkeyboard.h>
+
 using std::string;
 
 /* diagnostic routine:
@@ -790,6 +793,20 @@
 		}
 	}
 
+
+
+
+	SDL_Rect r;
+	r.h = disp_width / 30;
+	r.w = r.h;
+	r.x = disp_width - r.w;
+	r.y = r.h;
+	SDL_ANDROID_SetScreenKeyboardButtonPos(SDL_ANDROID_SCREENKEYBOARD_BUTTON_TEXT, &r);
+
+
+
+
+
 	dbg->important("Preparing display ...");
 	DBG_MESSAGE("simmain", "simgraph_init disp_width=%d, disp_height=%d, fullscreen=%d", disp_width, disp_height, fullscreen);
 	simgraph_init(disp_width, disp_height, fullscreen);
