diff --git a/gemrb/core/System/Logger/Win32Console.cpp b/gemrb/core/System/Logger/Win32Console.cpp
index 6bd9295..f2a5cff 100644
--- a/gemrb/core/System/Logger/Win32Console.cpp
+++ b/gemrb/core/System/Logger/Win32Console.cpp
@@ -16,6 +16,7 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#ifdef WIN32
 #include "System/Logger/Win32Console.h"
 
 #include "System/Logging.h"
@@ -77,3 +78,4 @@ Logger* createWin32ConsoleLogger()
 }
 
 }
+#endif
diff --git a/gemrb/core/System/Logger/Win32Console.h b/gemrb/core/System/Logger/Win32Console.h
index e28c685..600c157 100644
--- a/gemrb/core/System/Logger/Win32Console.h
+++ b/gemrb/core/System/Logger/Win32Console.h
@@ -22,6 +22,8 @@
 
 #include "System/Logger/Stdio.h"
 
+#ifdef WIN32
+
 namespace GemRB {
 
 class GEM_EXPORT Win32ConsoleLogger : public StdioLogger {
@@ -41,3 +43,4 @@ Logger* createWin32ConsoleLogger();
 }
 
 #endif
+#endif
diff --git a/gemrb/plugins/SDLVideo/SDL12Video.cpp b/gemrb/plugins/SDLVideo/SDL12Video.cpp
index ba533ba..240d6df 100644
--- a/gemrb/plugins/SDLVideo/SDL12Video.cpp
+++ b/gemrb/plugins/SDLVideo/SDL12Video.cpp
@@ -23,6 +23,9 @@
 #include "Game.h"
 #include "Interface.h"
 
+#if SDL_VERSION_ATLEAST(1,3,0)
+#else
+
 using namespace GemRB;
 
 SDL12VideoDriver::SDL12VideoDriver(void)
@@ -335,3 +338,5 @@ bool SDL12VideoDriver::SetSurfaceAlpha(SDL_Surface* surface, unsigned short alph
 GEMRB_PLUGIN(0xDBAAB50, "SDL Video Driver")
 PLUGIN_DRIVER(SDL12VideoDriver, "sdl")
 END_PLUGIN()
+
+#endif
diff --git a/gemrb/plugins/SDLVideo/SDL12Video.h b/gemrb/plugins/SDLVideo/SDL12Video.h
index c87a8f5..c2199b9 100644
--- a/gemrb/plugins/SDLVideo/SDL12Video.h
+++ b/gemrb/plugins/SDLVideo/SDL12Video.h
@@ -23,6 +23,9 @@
 
 #include "SDLVideo.h"
 
+#if SDL_VERSION_ATLEAST(1,3,0)
+#else
+
 namespace GemRB {
 
 class SDL12VideoDriver : public SDLVideoDriver {
@@ -63,3 +66,4 @@ private:
 }
 
 #endif
+#endif
diff --git a/gemrb/plugins/SDLVideo/SDLVideo.cpp b/gemrb/plugins/SDLVideo/SDLVideo.cpp
index f117374..0fa70c4 100644
--- a/gemrb/plugins/SDLVideo/SDLVideo.cpp
+++ b/gemrb/plugins/SDLVideo/SDLVideo.cpp
@@ -46,6 +46,17 @@ using namespace GemRB;
 #define SDL_SRCCOLORKEY SDL_TRUE
 #define SDL_SRCALPHA 0
 #define SDLK_SCROLLOCK SDLK_SCROLLLOCK
+// I just LOVE SDL backward compatibility, when it comes to keycodes
+#define SDLK_KP0 SDLK_KP_0
+#define SDLK_KP1 SDLK_KP_1
+#define SDLK_KP2 SDLK_KP_2
+#define SDLK_KP3 SDLK_KP_3
+#define SDLK_KP4 SDLK_KP_4
+#define SDLK_KP5 SDLK_KP_5
+#define SDLK_KP6 SDLK_KP_6
+#define SDLK_KP7 SDLK_KP_7
+#define SDLK_KP8 SDLK_KP_8
+#define SDLK_KP9 SDLK_KP_9
 #endif
 
 SDLVideoDriver::SDLVideoDriver(void)
